codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -1143,85 +831,43 @@ public long getTimeMillisAndRemove(T name, long defaultValue) {
-     * @return True if {@code lhs} == {@code rhs} according to {@code comparator}. False otherwise.
+     * This method is purposfully kept simple and returns {@link #size()} as the hash code.
+     * There are two compelling reasons for keeping {@link #hashCode()} simple:
+     *   1) It's difficult to get it right as the hash code mostly depends on the {@link Map}
+     *      implementation. Simply using {@link Map#hashCode()} doesn't work as for
+     *      example {@link TreeMap#hashCode()} does not fullfill the contract between {@link Object#hashCode()} and
+     *      {@link Object#equals(Object)} when it's used with a {@link Comparator} that is not consistent with
+     *      {@link Object#equals(Object)}.
+     *   2) The {@link #hashCode()} function doesn't appear to be important for {@link Headers} implementations. It's
+     *      soley there because we overode {@link Object#equals(Object)}, which makes it easier to use {@link Headers}
+     *      in tests, but also has little practical use outside of tests.
      */
-    private static <T> boolean equals(List<T> lhs, List<T> rhs, Comparator<? super T> comparator) {
-        final int lhsSize = lhs.size();
-        if (lhsSize != rhs.size()) {
-            return false;
-        }
-
-        // Don't use a TreeSet to do the comparison.  We want to force the comparator
-        // to be used instead of the object's equals()
-        Collections.sort(lhs, comparator);
-        Collections.sort(rhs, comparator);
-        for (int i = 0; i < lhsSize; ++i) {
-            if (comparator.compare(lhs.get(i), rhs.get(i)) != 0) {
-                return false;
-            }
-        }
-        return true;
-    }
-
     @Override
     public int hashCode() {
-        int result = 1;
-        for (T name : names()) {
-            result = HASH_CODE_PRIME * result + name.hashCode();
-            List<T> values = getAll(name);
-            Collections.sort(values, valueComparator);
-            for (int i = 0; i < values.size(); ++i) {
-                result = HASH_CODE_PRIME * result + hashCodeGenerator.generateHashCode(values.get(i));
-            }
-        }
-        return result;
+        return size();
     }
 
     @Override
     public String toString() {
         StringBuilder builder = new StringBuilder(getClass().getSimpleName()).append('[');
+        String separator = "";
         for (T name : names()) {
             List<T> values = getAll(name);
-            Collections.sort(values, valueComparator);
             for (int i = 0; i < values.size(); ++i) {
-                builder.append(name).append(": ").append(values.get(i)).append(", ");
+                builder.append(separator);
+                builder.append(name).append(": ").append(values.get(i));
             }
-        }
-        if (builder.length() > 2) { // remove the trailing ", "
-            builder.setLength(builder.length() - 2);
+            separator = ", ";
```

Comment:
It would be nice to avoid this duplicate assignment on every iteration of the loop.  Are you thinking JIT will optimize this away somehow?  If not consider unrolling the first iteration and all the rest to avoid this.